### PR TITLE
Add new step for waiting for presence of element by JSON path using polling interval

### PR DESCRIPTION
**Pull request by AlenaBudevich**
…erval  retrying retryTimes times stepsToExecute

### CODE REVIEW

File: vividus-util/src/main/java/org/vividus/util/wait/Waiter.java [hunk 1]
```diff
@@ -28,24 +29,35 @@
...
+        int counter = 0;
         do
         {
             value = valueProvider.get();
             Sleeper.sleep(pollingTimeoutMillis, TimeUnit.MILLISECONDS);
+            counter++;
         }
-        while (!stopCondition.test(value) && System.currentTimeMillis() <= endTime);
+        while (!stopCondition.test(value) && (durationInMillis > 0
+                ? System.currentTimeMillis() <= endTime : counter < pollingRetries));
```
Comment (valfirst) [inline vividus-util/src/main/java/org/vividus/util/wait/Waiter.java:60]: I think it makes sense to have 2 implementations of waiter:
- duration based
- retry times based

in this case logic won't be mixed up

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java [hunk 2]
```diff
@@ -400,7 +401,16 @@ void shouldWaitForJsonElement()
                 .thenReturn(createHttpResponse(JSON));
         when(httpTestContext.getJsonContext()).thenReturn(JSON);
         int retryTimes = 4;
-        jsonResponseValidationSteps.waitForJsonElement(STRING_PATH, Duration.ofSeconds(2), retryTimes, stepsToExecute);
+        if (withPollingInterval)
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java:404]: there should be 2 unit tests, such `if`-s can't be used in unit tests

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java [hunk 3]
```diff
@@ -308,9 +308,39 @@ public void waitForJsonFieldAppearance(String jsonPath, String resourceUrl, Dura
...
+     * <li>Stop step execution if HTTP response is not present or JSON element is found, otherwise
+     * sleep for the calculated part of specified polling interval and repeat actions from the start</li>
+     * </ul>
+     * @param jsonPath JSON path of element to find
+     * @param interval Duration of time to wait
+     * @param retryTimes Number of attempts
+     * @param stepsToExecute Steps to execute at each wait iteration
+     */
+    @When("I wait for presence of element by `$jsonPath` with polling interval `$interval` retrying $retryTimes times"
+            + "$stepsToExecute")
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java:332]:
```java
When I wait for presence of element by `$jsonPath` with `$pollingInterval` polling interval retrying $retryTimes times$stepsToExecute
```

Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java:332]: also please add docs for new step

File: vividus-util/src/main/java/org/vividus/util/wait/Waiter.java [hunk 4]
```diff
@@ -28,24 +29,35 @@
 {
     private final long durationInMillis;
     private final long pollingTimeoutMillis;
+    private final int pollingRetries;
```
Comment (valfirst) [inline vividus-util/src/main/java/org/vividus/util/wait/Waiter.java:32]: `pollingRetries` -> `retryTimes` - to keep naming consistency

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/plugins/pages/plugin-rest-api.adoc [hunk 5]
```diff
@@ -50,6 +50,40 @@ Scenario: Verify that the server supports TLSv1.3 and TLSv1.2 protocols
 Then server `vividus-test-site.herokuapp.com` supports secure protocols that contain `TLSv1.3,TLSv1.2`
 ----
 
+=== Wait for presence of element by JSON path
+
+==== *_Info_*
+
+Execute a specified amount of `retries` using `polling interval` of time until HTTP response body contains an element by the specified `JSON path`.
```
Comment (uarlouski) [inline docs/modules/plugins/pages/plugin-rest-api.adoc:57]: execute what? :)

Comment (AlenaBudevich) [reply docs/modules/plugins/pages/plugin-rest-api.adoc]: steps :)

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java [hunk 6]
```diff
@@ -407,6 +408,25 @@ void shouldWaitForJsonElement()
...
+                .thenReturn(createHttpResponse(HTML))
+                .thenReturn(createHttpResponse(OBJECT_PATH_RESULT))
+                .thenReturn(new HttpResponse())
+                .thenReturn(createHttpResponse(JSON));
+        when(httpTestContext.getJsonContext()).thenReturn(JSON);
+        int retryTimes = 4;
+        jsonResponseValidationSteps.waitForJsonElementWithPollingInterval(STRING_PATH,
+                Duration.ofSeconds(2), retryTimes, stepsToExecute);
+        verify(stepsToExecute, atLeast(retryTimes - 1)).execute(Optional.empty());
+        verify(stepsToExecute, atMost(retryTimes)).execute(Optional.empty());
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/JsonResponseValidationStepsTests.java:425]: I believe you can check exact number of invocations here

File: vividus-util/src/main/java/org/vividus/util/wait/DurationBasedWaiter.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+{
+    private final long durationInMillis;
+
+    public DurationBasedWaiter(WaitMode waitMode)
+    {
+        super(waitMode.calculatePollingTimeout(TimeUnit.MILLISECONDS));
+        durationInMillis = waitMode.getDuration().toMillis();
+    }
+
+    public <T, E extends Exception> T wait(FailableSupplier<T, E> valueProvider, Predicate<T> stopCondition) throws E
```
Comment (uarlouski) [inline vividus-util/src/main/java/org/vividus/util/wait/DurationBasedWaiter.java:35]: add override annotation for both waiters

**Review by uarlouski — APPROVED** [review]

**Review by valfirst — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Pull request merged**